src/learning/toolkit/cross_validation.rs

```diff
@@ -0,0 +1,235 @@
+//! Cross-validation
+
+use std::iter::Chain;
+use std::slice::Iter;
+use linalg::Matrix;
+use learning::SupModel;
+use learning::toolkit::cost_fn::*;
+use learning::toolkit::rand_utils::in_place_fisher_yates;
+
+// TODO: Support other input and output types.
+// TODO:
+// TODO: DON'T ALLOCATE DATA FOR EACH FOLD
+// TODO: See comment on copy_rows.
+// TODO:
+// TODO: Remove gradient from CostFunc and add a new trait
+// TODO: for differentiable cost functions.
+// TODO:
+// TODO: Clarify what happens when model.train is called multiple
+// TODO: times. This assumes that it throws away the old data and
+// TODO: trains a new model.
+/// Randomly splits the inputs into k 'folds'. For each fold a model
+/// is trained using all inputs except for that fold, and tested on the
+/// data in the fold. Returns the costs for each fold.
+pub fn k_fold_validate<M, C>(model: &mut M,
+                             inputs: &Matrix<f64>,
+                             targets: &Matrix<f64>,
+                             k: usize) -> Vec<f64>
+    where C: CostFunc<Matrix<f64>>,
+          M: SupModel<Matrix<f64>, Matrix<f64>>,
+{
+    assert_eq!(inputs.rows(), targets.rows());
+    let num_samples = inputs.rows();
+    let shuffled_indices = create_shuffled_indices(num_samples);
+    let folds = Folds::new(&shuffled_indices, k);
+
+    let mut costs: Vec<f64> = Vec::new();
+
+    for p in folds {
+        let train_inputs = copy_rows(&inputs, p.train_indices.clone());
+        let train_targets = copy_rows(&targets, p.train_indices.clone());
+        let test_inputs = copy_rows(&inputs, p.test_indices.clone());
+        let test_targets = copy_rows(&targets, p.test_indices.clone());
+
+        model.train(&train_inputs, &train_targets);
+        let outputs = model.predict(&test_inputs);
+        costs.push(C::cost(&outputs, &test_targets));
+    }
+
+    costs
+}
+
+// TODO: Use a preallocated buffer for each fold.
+fn copy_rows<'a, I>(mat: &Matrix<f64>,
+                    rows: I) -> Matrix<f64>
+    where I: ExactSizeIterator<Item=&'a usize>
+{
+    let num_rows = rows.len();
+    let mut data = vec![0f64; num_rows * mat.cols()];
+    let mut idx = 0;
+    for &row in rows {
+        for col in 0..mat.cols(){
+            data[idx] = mat[[row, col]];
+            idx += 1;
```

Comment:
We can make this loop more efficient ant remove the bound checks on both `mat` and `data`. Something like this:

```rust
let mut data = Vec::with_capacity(num_rows * mat.cols());

for &row in rows {
    unsafe {
        data.extend_from_slice(mat.get_row_unchecked(row));
    }
}
```